### ISSUE TITLE
Disable user VPN sessions when removed from OIDC provider

### ISSUE THREAD

**Issue by jamilbk** [CloudFire-LLC/cloudfire#585]
Add a periodic check to disconnect or disable a user's VPN if refreshing their token from the OIDC provider fails.

https://discourse.firez.one/t/google-sso-with-auto-disconnect/66

**Comment by jamilbk** [comment]
@gnavarroco AFAIK, OIDC doesn't yet provide a mechanism in the spec for this ([though they seem to be working on it](https://openid.net/specs/openid-risc-event-types-1_0-ID1.html#rfc.section.2.2)), so we'll need to come with a provider-specific workaround.

**Comment by mdp** [comment]
One option would be to request the 'offline' scope on the token, which would grant you a 'refresh token', then periodically perform a refresh and revoke the users access if the refresh token fails.

**Comment by jamilbk** [comment]
@mdp Good call! I think that's what most other OAuth client apps do. Not quite realtime but probably close enough. Thanks for the suggestion.

**Comment by gnavarroco** [comment]
> @gnavarroco AFAIK, OIDC doesn't yet provide a mechanism in the spec for this ([though they seem to be working on it](https://openid.net/specs/openid-risc-event-types-1_0-ID1.html#rfc.section.2.2)), so we'll need to come with a provider-specific workaround.

yeah i see :(, btw @mdp give a great idea

**Comment by princemaple** [comment]
Related reading: https://developers.google.com/identity/protocols/oauth2/web-server#httprest_1

tldr: Looks like it's the `access_type` that's relevant. Not really a scope.

Most helpful SO answer: https://stackoverflow.com/questions/30637984/what-does-offline-access-in-oauth-mean

**Comment by princemaple** [comment]
Okta does have an `offline_access` scope. https://developer.okta.com/docs/reference/api/oidc/#request-parameters-4

So we need docs to teach users about how to configure their openid providers if they want the ability to automatically disable users.

**Comment by jamilbk** [comment]
@gnavarroco This has been merged and should make it into the next release -- currently aiming for 0.4.0 next week sometime.

**Comment by gnavarroco** [comment]
🎉 thanks you guys!!

**Comment by jamilbk** [comment]
@gnavarroco May I ask which OIDC provider you’re using? I’ll make sure to test against it.
